### PR TITLE
Also retry reading connection responses

### PR DESCRIPTION
**Pull request by dasch**
We're still seeing errors from idle connections being used – hopefully this will fix the problem once and for all.

### CODE REVIEW

File: lib/kafka/connection.rb [hunk 1]
```diff
@@ -196,6 +181,21 @@ def read_response(response_class, notification)
       raise
     end
 
+    # Kafka brokers automatically close client connections after a period of
+    # inactivity. If this has happened, it's safe to re-open the connection
+    # and retry the request.
+    def with_retry
+      yield
+    rescue Errno::EPIPE, Errno::ECONNRESET, EOFError
```
Comment (tcrayford) [inline lib/kafka/connection.rb]: Are there equivalent errors for SSL sockets we should catch here?

Comment (dasch) [inline lib/kafka/connection.rb]: I'd assume that we'd get the same errors – my understanding is that the SSL "socket" is really just two normal sockets under the hood.

Comment (dasch) [inline lib/kafka/connection.rb]: And the errors would be at the TCP level.

Comment (tcrayford) [inline lib/kafka/connection.rb]: Having done some digging, seems accurate.

Comment (dasch) [inline lib/kafka/connection.rb]: :+1:

**Comment by ctrochalakis** [comment]
Just a quick thought,

We were doing some experiments on that and when the server closes the connection we see that the EOFError is always raised in `read_response()`.

In this patch you retry the `read_response()` part but perhaps you should retry sending the **request**. Trying reading a response (in a request/response cycle) doesn't make sense as the new connection has lost context.

Maybe retrying should be moved to upper levels where we have more context on what's the better thing to do. In the consumer, for example, we should retry sending the request.

**Comment by dasch** [comment]
@ctrochalakis good point. Let me think about this a bit. Currently we *do* retry requests, but not immediately and not always. I would only want to do retries at this low a level if it wasn't dangerous.

**Comment by dasch** [comment]
Closing this, it doesn't make sense.

**Pull request closed without merging**